products.js

```diff
@@ -0,0 +1,308 @@
+//       rule(type('priceish'), score(tagHasGoodCss)),
+//
+//       //not to be confused with amount off (minus sign, crossed off)
+//       rule(type('priceish'), score(notSavingsAmount)),
+//
+//       //return image with max score
+//       rule(type('priceish').max(), out('product-price'))
+//
+//     );
+//
+//     function contentFnodes(doc) {
+//         return rules.against(doc).get('product-price');
+//     }
+//
+//     return contentFnodes;
+// }
+
+/**
+ * Maintain state as we compare a series of DOMs, reporting the percent
+ * difference at the end.
+ */
+class DiffStats {
+    constructor(contentFnodes, feature) {
+        this.lengthOfExpectedTexts = 0;
+        this.lengthOfDiffs = 0;
+        this.feature = feature;
+        if (feature === 'image') {
+          this.contentFnodes = contentFnodes || tunedImageFnodes();
+        } else if (feature === 'title'){
+          this.contentFnodes = contentFnodes || tunedTitleFnodes();
+        } else {
+          this.contentFnodes = contentFnodes || tunedPriceFnodes();
+        }
+    }
+
+    compare(expectedDom, sourceDom) {
+        var expectedText;
+        var gotText;
+        if (this.feature === 'image') {
+          //compare images by src
+          expectedText = expectedDom.body.firstChild.src;
+          gotText = this.contentFnodes(sourceDom).map(fnode => fnode.element.src)[0];
+        } else if (this.feature === 'title') {
+          //compare innerHTML text of titles
+          expectedText = expectedDom.head.firstChild.innerHTML;
+          gotText = this.contentFnodes(sourceDom).map(fnode => fnode.element.innerHTML)[0];
+        } else if (this.feature === 'price') {
+          //strip whitespace and dollar sign if there is one when comparing price
+          expectedText = expectedDom.body.firstChild.textContent.replace('$', '').trim();
+          gotText = this.contentFnodes(sourceDom).map(fnode => fnode.element.textContent.replace('$', '').trim())[0];
+        }
+
+        this.lengthOfExpectedTexts++;
+        if(expectedText !== gotText) {
+          this.lengthOfDiffs++;
+        }
+
+        // Uncomment for debugging:
+        console.log(leven(expectedText, gotText), expectedText.length, leven(expectedText, gotText)/expectedText.length);
+        console.log('Got:\n' + gotText);
+        console.log('\nExpected:\n' + expectedText);
+    }
+
+    score() {
+        return this.lengthOfDiffs / this.lengthOfExpectedTexts * 100;
+    }
+}
+
+function createDict(item, sourceDom){
+  var nodesDict = JSON.parse(readFileSync(join(dirname(__dirname), 'fathom-products', 'product_classification_test_data', item, 'nodes.txt'), 'utf-8'));
+  var elems = sourceDom.getElementsByTagName('*');
+  console.log(sourceDom.getElementsByTagName('*').length, Object.keys(nodesDict).length);
+
+  //Match each node from the sourceDom to its css, and store in the global dict
+  for(let i = 0; i < Object.keys(nodesDict).length; i++){
+    currDict[elems[i]] = nodesDict[i.toString()];
```

Comment:
I'd be shocked if this worked, since currDict is a POJO (plain old JS object) rather than a Map. Does it work?

POJOs support only string keys, and I imagine many or all of the items in `elems` smoosh down to the same one (probably something like `[Object object[`). Or maybe somebody took the time to implement a unique toString() method on them. Even in that case, though, it's less efficient to make the trip through string construction every time rather than just hashing the object reference.